File: c/shaker_sort.c
#include <stdio.h>

void swap(int* arg1, int* arg2) {
    int temp = *arg1;
    *arg1 = *arg2;
    *arg2 = temp;
}

void shaker_sort(int arr[], int length) {
    length--;
    int changes = 1;
    while (changes) {
        changes = 0;
        for (int ind = 0; ind < length; ind++) {
            if (arr[ind] > arr[ind + 1]) {
                swap(&arr[ind], &arr[ind + 1]);
                changes = 1;
            }
        }
        if (changes == 0) break;
        changes = 0;
        for (int ind = length; ind > 0; ind--) {
            if (arr[ind] < arr[ind - 1]) {
                swap(&arr[ind], &arr[ind - 1]);
                changes = 1;
            }
        }
    } // no need to return the array
}

int main() {
    int example[] = {5, 1, 6, 3, 4};
    int length = sizeof(example) / 4;
    shaker_sort(example, length);
    for (int ind = 0; ind < length; ind++) {
        printf("%d ", example[ind]);
    }
    return 0;
}

File: javascript/shaker_sort.js
window.onload = function() {
    let example = [5, 1, 6, 3, 4];
    shaker_sort(example);
    let string = "";
    for (let ind = 0; ind < example.length; ind++) {
        string += example[ind] + ' ';
    }
    console.log(string);
}

function shaker_sort(arr) {
    let l = arr.length;
    let changes = true;
    while (changes) {
        changes = false;
        for (let ind = 0; ind < l; ind++) {
            if (arr[ind] > arr[ind + 1]) {
                [arr[ind], arr[ind + 1]] = [arr[ind + 1], arr[ind]]
                changes = true;
            }
        }
        if (!changes) break;
        changes = false;
        for (let ind = l; ind > 0; ind--) {
            if (arr[ind] < arr[ind - 1]) {
                [arr[ind], arr[ind - 1]] = [arr[ind - 1], arr[ind]]
                changes = true;
            }
        }
    } // no need to return the array
}